Source/WebCore/rendering/style/TextSizeAdjustment.h

```diff
@@ -28,24 +28,34 @@ namespace WebCore {
 
 class RenderStyle;
 
-enum TextSizeAdjustmentType { AutoTextSizeAdjustment = -1, NoTextSizeAdjustment = -2 };
-
 class TextSizeAdjustment {
 public:
-    constexpr TextSizeAdjustment() : m_value(AutoTextSizeAdjustment) { }
-    constexpr TextSizeAdjustment(float value) : m_value(value) { }
-
-    constexpr float percentage() const { return m_value; }
-    constexpr float multiplier() const { return m_value / 100; }
-
-    constexpr bool isAuto() const { return m_value == AutoTextSizeAdjustment; }
-    constexpr bool isNone() const { return m_value == NoTextSizeAdjustment; }
-    constexpr bool isPercentage() const { return m_value >= 0; }
+    enum class Type : bool { None, Auto };
+
+    constexpr TextSizeAdjustment() : m_value(TextSizeAdjustment::Type::Auto) { }
+    constexpr TextSizeAdjustment(float value) : m_value(value) { ASSERT(value >= 0); }
+    constexpr TextSizeAdjustment(TextSizeAdjustment::Type value) : m_value(value) { }
+
+    constexpr float percentage() const { return std::get<float>(m_value); }
+    constexpr float multiplier() const { return std::get<float>(m_value) / 100; }
+
+    constexpr bool isAuto() const
+    {
+        return std::holds_alternative<TextSizeAdjustment::Type>(m_value) && std::get<TextSizeAdjustment::Type>(m_value) == TextSizeAdjustment::Type::Auto;
+    }
+    constexpr bool isNone() const
+    {
+        return std::holds_alternative<TextSizeAdjustment::Type>(m_value) && std::get<TextSizeAdjustment::Type>(m_value) == TextSizeAdjustment::Type::None;
+    }
+    constexpr bool isPercentage() const
+    {
+        return std::holds_alternative<float>(m_value) && std::get<float>(m_value) >= 0;
+    }
 
     friend constexpr bool operator==(TextSizeAdjustment, TextSizeAdjustment) = default;
 
 private:
-    float m_value;
+    std::variant<float, TextSizeAdjustment::Type> m_value;
```

Comment:
This makes this bigger than 32 bits. That does not seem like a good tradeoff.

Reply:
Instead we can still use special values for auto and none, but have those special values be private `constexpr` class members. And not use `std::variant`.

Reply:
As noted in the commit log message, `dump-class-layout` reports that this PR changes the `WebCore::TextSizeAdjustment` class from 4 bytes to 8 bytes.

Due to existing padding of 4 bytes after `WebCore::TextSizeAdjustment` in `WebCore::StyleRareInheritedData`, this change does not increase the size of `StyleRareInheritedData`.

```
$ xcrun python3 ./Tools/Scripts/dump-class-layout -a arm64e -b /var/build -c Release WebCore WebCore::TextSizeAdjustment
  +0 <  8> WebCore::TextSizeAdjustment
  +0 <  8>     std::__1::variant<float, WebCore::TextSizeAdjustment::Type> m_value
  +0 <  1>         std::__1::__sfinae_ctor_base<true, true> std::__1::__sfinae_ctor_base<true, true>
  +0 <  1>         std::__1::__sfinae_assign_base<true, true> std::__1::__sfinae_assign_base<true, true>
  +0 <  8>         std::__1::__variant_detail::__impl<float, WebCore::TextSizeAdjustment::Type> __impl_
  +0 <  8>             std::__1::__variant_detail::__copy_assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable> std::__1::__variant_detail::__copy_assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable>
  +0 <  8>                 std::__1::__variant_detail::__move_assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable> std::__1::__variant_detail::__move_assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable>
  +0 <  8>                     std::__1::__variant_detail::__assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type> > std::__1::__variant_detail::__assignment<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type> >
  +0 <  8>                         std::__1::__variant_detail::__copy_constructor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable> std::__1::__variant_detail::__copy_constructor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable>
  +0 <  8>                             std::__1::__variant_detail::__move_constructor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable> std::__1::__variant_detail::__move_constructor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable>
  +0 <  8>                                 std::__1::__variant_detail::__ctor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type> > std::__1::__variant_detail::__ctor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type> >
  +0 <  8>                                     std::__1::__variant_detail::__dtor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable> std::__1::__variant_detail::__dtor<std::__1::__variant_detail::__traits<float, WebCore::TextSizeAdjustment::Type>, std::__1::__variant_detail::_Trait::_TriviallyAvailable>
  +0 <  8>                                         std::__1::__variant_detail::__base<std::__1::__variant_detail::_Trait::_TriviallyAvailable, float, WebCore::TextSizeAdjustment::Type> std::__1::__variant_detail::__base<std::__1::__variant_detail::_Trait::_TriviallyAvailable, float, WebCore::TextSizeAdjustment::Type>
  +0 <  4>                                           std::__1::__variant_detail::__union<std::__1::__variant_detail::_Trait::_TriviallyAvailable, 0, float, WebCore::TextSizeAdjustment::Type> __data
  +4 <  4>                                           __index_t __index
Total byte size: 8
Total pad bytes: 0
```

Reply:
Also note that the `TextSizeAdjustment::Type` enum class is now uses a `bool` for storage.

Reply:
Oops, you said 32 _bits_, not 32 _bytes_.  I misread that earlier.